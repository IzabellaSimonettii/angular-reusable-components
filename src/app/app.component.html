<section>
  <button customButton color="primary">
    primary
  </button>

  <button customButton color="secondary">
    secondary
  </button>

  <button customButton outline>
    outline
  </button>

  <button customButton outline>
    outline icon left
    <img src="/assets/send.svg" alt="" icon />
  </button>

  <button customButton color="primary" iconPosition="right">
    icon right
    <img src="/assets/send.svg" alt="" icon />
  </button>

  <button customButton disabled>
    disabled
  </button>
</section>
