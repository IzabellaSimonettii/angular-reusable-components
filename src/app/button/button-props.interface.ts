export default interface IButtonProps {
  iconPosition: string;
  color: string;
  outline: boolean;
}
