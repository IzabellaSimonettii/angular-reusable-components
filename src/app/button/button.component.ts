import {
  Component,
  OnInit,
  ViewEncapsulation,
  HostBinding,
  Input,
} from '@angular/core';

import IButtonProps from './button-props.interface';


@Component({
  selector: 'button[customButton], a[customButton]',
  templateUrl: './button.component.html',
  styleUrls: ['./button.component.scss'],
  encapsulation: ViewEncapsulation.None,
})
export class ButtonComponent implements OnInit , IButtonProps {
  @Input()
  iconPosition: 'left' | 'right' = 'left';

  @Input()
  color: 'primary' | 'secondary';

  @Input()
  set outline(value: boolean) {
    this._outline = value !== null && `${value}` !== 'false';
  }

  @HostBinding('class.custom-button')
  _customButtom = true;

  @HostBinding('class.custom-button--outline')
  _outline = false;

  @HostBinding('class.custom-button--primary')
  get light(): boolean {
    return this.color === 'primary'
  }

  @HostBinding('class.custom-button--secondary')
  get dark(): boolean {
    return this.color === 'secondary'
  }


  constructor() {}

  ngOnInit(): void {}
}
