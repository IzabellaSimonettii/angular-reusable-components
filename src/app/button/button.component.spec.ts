import {
  SpectatorHost,
  createHostFactory,
} from '@ngneat/spectator/jest';
import { axe } from 'jest-axe';

import { APP_BASE_HREF } from '@angular/common';
import { AppModule } from '../app.module';
import { ButtonComponent } from './button.component';


const cases = {
  simple: `<button customButton>Hello World!</button>`,
  secondary: `<button customButton color="secondary">Hello World!</button>`,
  outline: `<button customButton outline>Hello World!</button>`,
};

describe('ButtonComponent', () => {
  let subject: SpectatorHost<ButtonComponent>;
  const createHost = createHostFactory({
    component: ButtonComponent,
    declareComponent: false,
    imports: [ AppModule ],
    providers: [{provide: APP_BASE_HREF, useValue: '/'}],
  });

  it('should create instance', () => {
    subject = createHost(cases.simple);
    expect(subject.component).toBeTruthy();
  });

  it('should create the raw component', () => {
    subject = createHost(cases.simple);
    expect(subject.hostFixture).toMatchSnapshot();
  });

  it('should have class custom-button--secondary', () => {
    subject = createHost(cases.secondary);
    expect(subject.hostFixture).toMatchSnapshot();
  });

  it('should have class custom-button--outline', () => {
    subject = createHost(cases.outline);
    expect(subject.hostFixture).toMatchSnapshot();
  });

  it('should verify component accessibility', async () => {
    subject = createHost(cases.simple);
    expect(await axe(subject.hostFixture.nativeElement)).toHaveNoViolations();
  });
});
