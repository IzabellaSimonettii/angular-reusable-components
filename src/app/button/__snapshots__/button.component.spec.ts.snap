// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonComponent should create the raw component 1`] = `
<ng-component>
  <button
    class="custom-button"
    custombutton=""
  >
    <span
      class="custom-button__wrapper custom-button__wrapper__icon-left"
    >
      Hello World!
    </span>
  </button>
</ng-component>
`;

exports[`ButtonComponent should have class custom-button--outline 1`] = `
<ng-component>
  <button
    class="custom-button custom-button--outline"
    custombutton=""
    outline=""
  >
    <span
      class="custom-button__wrapper custom-button__wrapper__icon-left"
    >
      Hello World!
    </span>
  </button>
</ng-component>
`;

exports[`ButtonComponent should have class custom-button--secondary 1`] = `
<ng-component>
  <button
    class="custom-button custom-button--secondary"
    color="secondary"
    custombutton=""
  >
    <span
      class="custom-button__wrapper custom-button__wrapper__icon-left"
    >
      Hello World!
    </span>
  </button>
</ng-component>
`;
